src/js/tracing/reactnativenavigation.ts

```diff
@@ -0,0 +1,208 @@
+import { Transaction as TransactionType } from "@sentry/types";
+import { logger } from "@sentry/utils";
+import { EmitterSubscription } from "react-native";
+
+import { BeforeNavigate } from "./reactnativetracing";
+import {
+  RoutingInstrumentation,
+  TransactionCreator,
+} from "./routingInstrumentation";
+import { getBlankTransactionContext } from "./utils";
+
+interface ReactNativeNavigationOptions {
+  routeChangeTimeoutMs: number;
+}
+
+const defaultOptions: ReactNativeNavigationOptions = {
+  routeChangeTimeoutMs: 1000,
+};
+
+interface ComponentEvent {
+  componentId: string;
+}
+
+type ComponentType =
+  | "Component"
+  | "TopBarTitle"
+  | "TopBarBackground"
+  | "TopBarButton";
+
+export interface ComponentWillAppearEvent extends ComponentEvent {
+  componentName: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  passProps?: Record<any, any>;
+  componentType: ComponentType;
+}
+
+export interface EventSubscription {
+  remove(): void;
+}
+
+export interface EventsRegistry {
+  registerComponentWillAppearListener(
+    callback: (event: ComponentWillAppearEvent) => void
+  ): EmitterSubscription;
+  registerCommandListener(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    callback: (name: string, params: any) => void
+  ): EventSubscription;
+}
+
+export interface NavigationDelegate {
+  events: () => EventsRegistry;
+}
+
+/**
+ * Instrumentation for React Native Navigation. See docs or sample app for usage.
+ *
+ * How this works:
+ * - `_onCommand` is called every time a commands happens and sets an IdleTransaction on the scope without any route context.
+ * - `_onComponentWillAppear` is then called AFTER the state change happens due to a dispatch and sets the route context onto the active transaction.
+ * - If `_onComponentWillAppear` isn't called within `STATE_CHANGE_TIMEOUT_DURATION` of the dispatch, then the transaction is not sampled and finished.
```

Comment:
Where is `STATE_CHANGE_TIMEOUT_DURATION` defined? Also where does this logic happen?

Reply:
should be `options.routeChangeTimeoutMs`